### ISSUE TITLE
[priority] Memory leak introduced after v0.18

### ISSUE THREAD

**Issue by b-ma** [ircam-ismm/node-web-audio-api#118]
Nodes are not garbage collected as they should, clearly visible with granular scrub example...

Do not use v0.19

**Comment by b-ma** [comment]
This seems to be the faulty commit: https://github.com/ircam-ismm/node-web-audio-api/commit/8d819d49f8fae662b73a1059d0054ab36a4904da

**Comment by b-ma** [comment]
still problematic on RPi in 0.21.0 (only half the problem seems to be solved...)

**Comment by b-ma** [comment]
- [ ] bisect between 0.18 and 0.19 to hopefully find another faulty commit
- [ ] AudioBuffer could be refactored to raw object maybe same issue as for AudioParam (this was there before 0.18 but not wrapped in a facade...)

**Comment by b-ma** [comment]
First step is to cross-compile `node-web-audio-api.linux-arm64-gnu.node` target in a docker image as building on the RPi itself take dozens of minutes, which makes the thing impossible to debug and test in practice

**Comment by orottier** [comment]
Do you see the memory leak only in the RPi? We would expect it to be present on all platforms, no?

You can use the hidden method `AudioContext::run_diagnostics` to print the current audio graph to see if nodes are not reclaimed properly. The issue might also be somewhere else of course (unhandled events, etc)

**Comment by b-ma** [comment]
> Do you see the memory leak only in the RPi? We would expect it to be present on all platforms, no?

Yup you are right, it's just that it creates really audible cracks on RPi (with CPU going up to 50% instead of 10-15%) while it sounds ok on my machine. But after rechecking I can indeed see the CPU going from ~2.6 to almost 9% too... while with version 0.18 it does not go beyond 3%

> You can use the hidden method AudioContext::run_diagnostics to print the current audio graph to see if nodes are not reclaimed properly. The issue might also be somewhere else of course (unhandled events, etc)

Thank for the hint I will have a look. My guess is that this to is due to something in how the bindings are made which prevent GC and makes the graph grow and grow. My best suspects right now are:
- Internally created classes wrapped in facade and then used as attributes, cf. previous issue with `AudioParam` which seemed to create the exact same problem
- The way event callbacks are handled (maybe the first simple to test as a first step)

I'm a bit nervous with that stuff considering all the work and refactoring that has been made since version 0.18...

**Comment by b-ma** [comment]
Some little info, no idea why and how to fix that but that's a first step... AudioBufferSourceNode are not collected in Scavenge, only GainNoes are...:

```
[20747:0x130008000]     6339 ms: Scavenge 9.2 (12.7) -> 8.7 (17.7) MB, pooled: 0 MB, 0.96 / 0.04 ms  (average mu = 1.000, current mu = 1.000) task;
NAPI: NapiGainNode dropped
NAPI: NapiGainNode dropped
// ....
[20747:0x130008000]     8119 ms: Mark-Compact (reduce) 10.1 (17.7) -> 5.0 (10.0) MB, pooled: 0 MB, 3.04 / 1.00 ms  (+ 1.4 ms in 0 steps since start of marking, biggest step 0.0 ms, walltime since start of marking 6 ms) (average mu = 0.999, current mu = 0.999) finalize incremental marking via task; GC in old space requested
NAPI: NapiAudioBufferSourceNode dropped
NAPI: NapiAudioBufferSourceNode dropped
NAPI: NapiAudioBufferSourceNode dropped
```

to be continued...

**Comment by b-ma** [comment]
Same behaviour if I replace the AudioBufferSourceNode with an OscillatorNode, interesting...

**Comment by b-ma** [comment]
-Next try remove all `AudioScheduledSourceNode#ended` event logic

**Comment by b-ma** [comment]
Fixed w/ #130
